src/Raven.Client/Util/WeakReferencingTimer.cs

```diff
@@ -0,0 +1,64 @@
+﻿using System;
+using System.Threading;
+
+namespace Raven.Client.Util;
+
+internal class WeakReferencingTimer : IDisposable
+{
+    internal delegate void WeakReferencingTimerCallback(object state);
+
+    private readonly Timer _timer;
+
+    public WeakReferencingTimer(WeakReferencingTimerCallback callback, object state, TimeSpan dueTime, TimeSpan period)
+    {
+        if (callback == null)
+            throw new ArgumentNullException(nameof(callback));
+
+        var internalTimerState = new TimerState
+        {
+            State = state is not null ? new WeakReference<object>(state) : null,
+            Callback = new WeakReference<WeakReferencingTimerCallback>(callback)
+        };
+
+        _timer = new Timer(StaticCallback, internalTimerState, dueTime, period);
+
+        internalTimerState.Timer = _timer;
+    }
+
+    private static void StaticCallback(object state)
+    {
+        var timerState = (TimerState)state;
+        object stateObj = null;
+
+        if (timerState.State is not null && timerState.State.TryGetTarget(out stateObj) == false)
+        {
+            timerState.Timer.Dispose();
+            return;
+        }
+
+        if (timerState.Callback.TryGetTarget(out var callback) == false)
+        {
+            timerState.Timer.Dispose();
+            return;
+        }
+
+        callback(stateObj);
+    }
+
+    private class TimerState
+    {
+        public WeakReference<object> State;
+        public WeakReference<WeakReferencingTimerCallback> Callback;
```

Comment:
That is most likely a problem. Who is going to be holding a reference to the callback in this case?
Better to associate the lifetime using `WeakConditionalTable` instead.

Reply:
We hold the reference to the callback - via `WeakReference`. What would be the benefit of using `ConditionalWeakTable` here? I don't see the need of using it.